### PR TITLE
Aho-Corasick algorithm editorial

### PR DESCRIPTION
**Pull request by brebenelmihnea**
_Place an "x" in the corresponding checkbox if it is done or does not apply to this pull request._

- [x] I have tested my code.
- [x] I have added my solution according to the steps [here](https://usaco.guide/general/adding-solution#steps).
- [x] I have followed the code conventions mentioned [here](https://usaco.guide/general/adding-solution/#code-conventions).
  - I understand that if it is clear that I have not attempted to follow these conventions, my PR will be closed.
  - If changes are requested, I will re-request a review after addressing them.
- [ ] I have linked this PR to any issues that it closes.

### CODE REVIEW

**Comment by brebenelmihnea** [comment]
image

**Comment by brebenelmihnea** [comment]
The images in advanced section don't have an assets folder. Should I create one and move all of them there ?

File: content/6_Advanced/String_Search.mdx [hunk 1]
```diff
@@ -414,6 +407,120 @@ Build the entire trie first, and then run a _BFS_ to construct the fail array.
...
+
+While processing the string $S$ the algorithm maintains the current node in the trie such that word formed in the node is equal to the longest suffix ending in $i$. For example, when
+transitioning from $i$ to $i+1$ in $S$ there only are two choices:
+1. If $node$ does have an outgoing edge with letter $S_{i+1}$, then move down the edge.
+2. Otherwise, follow the failure link of $S_i$ and transition to letter $S_{i+1}$ from there.
+
+The image below depicts how the structure looks like for the wordlist: a, ag, c, caa, gag, gc.
+
+<center>
+![Trie](./Trie.png)
```
Comment (SansPapyrus683) [inline content/6_Advanced/String_Search.mdx:429]: wait this image is actually amazing, good job!

File: content/6_Advanced/String_Search.mdx [hunk 2]
```diff
@@ -414,6 +407,120 @@ Build the entire trie first, and then run a _BFS_ to construct the fail array.
...
+int fail[NMAX], seen[NMAX], ans[NMAX];
+
+/* leaf[node] stores the indices of the words ending in node */
+vector<int> leaf[NMAX];
+vector<int> g[NMAX];
+
+/* Add a word to the trie */
+void add_word(const string &word, const int &idx) {
+	int node = 1;
+	for (auto ch : word) {
```
Comment (SansPapyrus683) [inline content/6_Advanced/String_Search.mdx:462]: plz don't use auto

File: content/6_Advanced/String_Search.mdx [hunk 3]
```diff
@@ -414,6 +407,129 @@ Build the entire trie first, and then run a _BFS_ to construct the fail array.
...
+		if (trie[node][ch - 'a'] == 0) { trie[node][ch - 'a'] = ++nodes; }
+		node = trie[node][ch - 'a'];
+	}
+	leaf[node].push_back(idx);
+}
+
+/** BFS for building the failure and suffix links */
```
Comment (SansPapyrus683) [inline content/6_Advanced/String_Search.mdx:469]:
```suggestion
/** BFS to building the failure and suffix links */
```

File: content/6_Advanced/String_Search.mdx [hunk 4]
```diff
@@ -414,6 +407,129 @@ Build the entire trie first, and then run a _BFS_ to construct the fail array.
...
+using namespace std;
+
+const int NMAX = 6e5;
+const int SIGMA = 26;
+
+int n;
+string s;
+int nodes = 1;
+int trie[NMAX][SIGMA];
+int fail[NMAX], seen[NMAX], ans[NMAX];
```
Comment (SansPapyrus683) [inline content/6_Advanced/String_Search.mdx:453]: comments for `leaf` but not any of these?

**Review by SansPapyrus683 — CHANGES_REQUESTED** [review]
W

File: content/6_Advanced/String_Search.mdx (hunk 2 above)
Comment (brebenelmihnea) [reply content/6_Advanced/String_Search.mdx:462]: You mean `for(char ch : word)` ?

File: content/6_Advanced/String_Search.mdx (hunk 4 above)
Comment (brebenelmihnea) [reply content/6_Advanced/String_Search.mdx]: Done

File: content/6_Advanced/String_Search.mdx (hunk 2 above)
Comment (brebenelmihnea) [reply content/6_Advanced/String_Search.mdx]: Done

Comment (SansPapyrus683) [reply content/6_Advanced/String_Search.mdx]: yeah

File: content/6_Advanced/String_Search.mdx [hunk 5]
```diff
@@ -414,6 +407,132 @@ Build the entire trie first, and then run a _BFS_ to construct the fail array.
...
+
+<LanguageSection>
+<CPPSection>
+
+```cpp
+#include <bits/stdc++.h>
+
+using namespace std;
+
+const int NMAX = 6e5;
```
Comment (SansPapyrus683) [inline content/6_Advanced/String_Search.mdx:446]: plz rename to `MAX_N`?

File: content/6_Advanced/String_Search.mdx [hunk 6]
```diff
@@ -414,6 +407,132 @@ Build the entire trie first, and then run a _BFS_ to construct the fail array.
...
+const int SIGMA = 26;
+
+int n;
+string s;
+// The number of nodes in trie
+int nodes = 1;
```
Comment (SansPapyrus683) [inline content/6_Advanced/String_Search.mdx:452]:
```suggestion
int nodes;  // The number of nodes in our trie
```

Comment (brebenelmihnea) [reply content/6_Advanced/String_Search.mdx:452]: ok

File: content/6_Advanced/String_Search.mdx (hunk 5 above)
Comment (brebenelmihnea) [reply content/6_Advanced/String_Search.mdx]: done

**Review by SansPapyrus683 — APPROVED** [review]
screenshot of _updated_ code AC-ing and we're done!

**Review by ryanchou-dev — APPROVED** [review]
great work with the explanation!

**Comment by brebenelmihnea** [comment]
@SansPapyrus683 
image

**Pull request merged**